Add retries to InstallVisualStudio.ps1 script

dougbu opened the pull request:
- short-term attempt to reduce CI problems in this area; see dotnet/core-eng#5760
- add `-ErrorAction Continue` to the `Start-Process` command line to enable retries

nits & debugging:
- display and interpret any exit code from the VS installer
- display VS installation logs
- remove unnecessary `[CmdletBinding]` attribute

dougbu committed:
Add retries to InstallVisualStudio.ps1 script
- short-term attempt to reduce CI problems in this area; see dotnet/core-eng#5760
- add `-ErrorAction Continue` to the `Start-Process` command line to enable retries

nits & debugging:
- display and interpret any exit code from the VS installer
- display VS installation logs
- remove unnecessary `[CmdletBinding]` attribute

dougbu committed:
Fix typo

dougbu committed:
And another typo...

dougbu commented:
Well, this worked on its first attempt @aspnet/build. VS installation added from 16 seconds (newest agents) to 9 minutes (had to install BT from scratch because Preview version was installed). Surprisingly, I didn't see any actual retries in this build.

natemcmaster commented on eng/scripts/InstallVisualStudio.ps1:
```diff
@@ -103,10 +102,60 @@ Write-Host "Installing Visual Studio 2019 $Edition" -f Magenta
+    } else {
+        # https://docs.microsoft.com/en-us/visualstudio/install/use-command-line-parameters-to-install-visual-studio#error-codes
+        if ($process.ExitCode -eq 3010) {
+            Write-Error "Installation requires restart to finish the VS update."
```
`Write-Error` is fatal (equivalent to 'throw Exception') when  `$ErrorActionPreference = 'Stop'`. Use `write-warning` or `write-host -f Red "..."` instead

dougbu replied:
Changed the preference [above](https://github.com/aspnet/AspNetCore/pull/9300/files#diff-27fbe4ead004bbb25de7d0b254c584fdR105) instead

natemcmaster replied:
Changing $ErrorActionPreference in the middle of a script seems to me like a bad pattern. Most good powershell code I've read leaves `$ErrorActionPreference = 'Stop'` throughout the whole script, only using `-ErrorAction continue` when appropriate. But I've also never seen `Write-Error -ErrorAction continue` because you can just do `Write-warning`.

dougbu replied:
:ok: I'll add the requisite `-ErrorAction continue` parameters and switch to `Write-Warning`

dougbu committed:
Address PR comments

dougbu commented on eng/scripts/InstallVisualStudio.ps1:
```diff
@@ -103,10 +102,60 @@ Write-Host "Installing Visual Studio 2019 $Edition" -f Magenta
+
+        $bootstrapperLogs = Get-ChildItem $env:Temp\dd_bootstrapper_*.log |Sort-Object CreationTime
+        if ($bootstrapperLogs.Count -ne 0) {
+            $bootstrapperLog = $bootstrapperLogs[$bootstrapperLogs.Count - 1]
```
Leaving as-is. `-Top` was just introduced in PowerShell 6 and I like having the filenames split up the logs.

natemcmaster replied:
Challenge accepted! 

```ps1
Get-ChildItem $env:Temp\dd_bootstrapper_*.log | sort CreationTime -Descending | select -first 1  | % { Write-Host $_; Get-Content $_ | Write-Host }
```

dougbu commented:
Will update the branch again. But, this PR is in an interesting state: The AspNetCore-ci run is reporting a test failure from the first of two attempts. @aspnet/build is this an AzDO issue or basically irrelevant?

natemcmaster commented:
Thanks.

Looks to me like more GitHub/AzDO integration issues. PR checks look fine.

dougbu committed:
Address PR comments
- do not reset `$ErrorActionPreference` (add `-ErrorAction Continue`)
- compact PowerShell that displays VS log files
- use `Write-Host -ForegroundColor Red` instead of `Write-Error`
  - avoid `Write-Error`'s extra lines of output

nits:
- spell out `-ForegroundColor Red`
- add "Error: " prefix to error messages
- remove unnecessary empty strings passed to `Write-Host`

dougbu commented:
:up::date: (also rebased on 'master')

dougbu committed:
Typo

dougbu commented:
@aspnet/build do we have an issue tracking this short-term mitigation?
